### PR TITLE
[wip] gpu should know packet layout

### PR DESCRIPTION
**Pull request by aeyakovenko**
@sakridge

### CODE REVIEW

File: src/transaction.rs [hunk 1]
```diff
@@ -133,4 +139,21 @@ mod tests {
...
+        let mut tr = Transaction::new(&keypair1, pubkey1, 42, zero);
+        tr.sign(&keypair1);
+        return tr;
+    }
+    #[test]
+    fn test_layout() {
+        let tr = test_tx();
+        let sign_data = tr.get_sign_data();
+        let tx = serialize(&tr).unwrap();
+        assert_match!(tx.find(sign_data), Some(64));
```
Comment (aeyakovenko) [inline src/transaction.rs]: @sakridge just assert the layout directly

File: src/accountant_skel.rs [hunk 2]
```diff
@@ -168,6 +197,18 @@ impl<W: Write + Send + 'static> AccountantSkel<W> {
...
+
+#[cfg(test)]
+mod tests {
+    use transaction::tests::test_tx;
+    #[test]
+    fn test_layout() {
+        let tr = test_tx();
+        let tx = serialize(&tr).unwrap();
+        let packet = serialize(Request::Transaction(tr));
+        assert_match!(packet.find(tx), Some(0));
```
Comment (aeyakovenko) [inline src/accountant_skel.rs]: @sakridge just assert the layout directly

File: src/gpu.rs [hunk 3]
```diff
@@ -0,0 +1,62 @@
+use libc::size_t;
+use std::mem::size_of;
+use streamer::SharedPackets;
+
+#[link(name = "cuda_verify_ed25519")]
+extern "C" {
+    fn ed25519_verify_many(
+        packets: *const u8,
```
Comment (aeyakovenko) [inline src/gpu.rs]: @sakridge this should be an iovec call, so vector of buffers all with the same layout
[sig, payload, up to 256 bytes][usize]

**Comment by garious** [comment]
Can you rebase?

**Comment by garious** [comment]
Please help me set a good example for future contributors: https://gist.github.com/matthewhudson/1475276

**Comment by aeyakovenko** [comment]
Sure!  I think we should squash things before merge.  At mesoshere, we did [WIP] tags on PRs while they were heavily in development, and then [READY] for review after squash + commit message.  It worked pretty well

File: src/transaction.rs [hunk 4]
```diff
@@ -29,11 +25,13 @@ impl Transaction {
         let from = from_keypair.pubkey();
         let plan = Plan::Pay(Payment { tokens, to });
         let mut tr = Transaction {
-            from,
-            plan,
-            tokens,
-            last_id,
             sig: Signature::default(),
+            data: Signed {
```
Comment (garious) [inline src/transaction.rs]: Can you rename that from Signed to TransactionData? Good call on grouping that.

File: src/accountant_skel.rs [hunk 5]
```diff
@@ -59,9 +60,9 @@ impl<W: Write + Send + 'static> AccountantSkel<W> {
         self.last_id
     }
 
-    pub fn process_request(self: &mut Self, msg: Request) -> Option<Response> {
+    pub fn process_request(self: &mut Self, msg: Request, verify: u8) -> Option<Response> {
         match msg {
-            Request::Transaction(tr) => {
+            Request::Transaction(tr) if verify != 0 => {
```
Comment (garious) [inline src/accountant_skel.rs]: Should that be a bool?

File: src/gpu.rs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        signed_message_offset: u32,
+        signed_message_len_offset: u32,
+        out: *mut u8, //combined length of all the items in vecs
+    ) -> u32;
+}
+
+pub fn ecdsa_verify(batches: &Vec<SharedPackets>) -> Vec<Vec<u8>> {
+    let mut out = Vec::new();
+    let mut elems = Vec::new();
+    let mut locks = Vec::new();
```
Comment (garious) [inline src/gpu.rs]: Needs a better name

File: src/gpu.rs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    let mut rvs = Vec::new();
+    for packets in batches {
+        locks.push(packets.read().unwrap());
```
Comment (garious) [inline src/gpu.rs]: map()?

File: src/gpu.rs [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        );
+        if res != 0 {}
+    }
+    println!("done verify");
+    let mut num = 0;
+    for vs in rvs {
+        for v in &mut vs {
+            *v = out[num]; 
+            num += 1;
+        }
```
Comment (garious) [inline src/gpu.rs]: map() and a += vs.len()?

File: src/transaction.rs [hunk 9]
```diff
@@ -63,13 +70,25 @@ impl Transaction {
...
 
     pub fn verify(&self) -> bool {
-        self.sig.verify(&self.from, &self.get_sign_data()) && self.plan.verify(self.tokens)
+        self.sig.verify(&self.from, &self.get_sign_data())
+            && self.data.plan.verify(self.data.tokens)
     }
 }
 
+#[cfg(test)]
+pub fn test_tx() -> Transaction {
```
Comment (garious) [inline src/transaction.rs]: Move this into mod tests below and give it a name that's a verb. Just a create_ prefix would be fine.

Comment (aeyakovenko) [reply src/transaction.rs]: that would make mod tests public

File: src/accountant.rs [hunk 10]
```diff
@@ -92,11 +94,11 @@ impl Accountant {
     }
 
     pub fn process_transaction(self: &mut Self, tr: Transaction) -> Result<()> {
-        if !tr.verify() {
+        if !tr.plan_verify() {
```
Comment (garious) [inline src/accountant.rs]: Method names need to be either `<verb>` or `<verb>_<subject>`. Can you change this to `verify_plan()`?

File: src/accountant_skel.rs [hunk 11]
```diff
@@ -92,43 +97,66 @@ impl<W: Write + Send + 'static> AccountantSkel<W> {
...
-                            .resize((num + 1) * 2, streamer::Response::default());
+        let batches = r_reader.recv_timeout(timer)?;
+        for (msgs, verifies) in batches.0.iter().zip(batches.1.iter()) {
+            let msgs_ = msgs.clone();
+            let rsps = streamer::allocate(response_recycler);
+            let rsps_ = rsps.clone();
+            {
+                let mut num = 0;
+                let mut ursps = rsps.write().unwrap();
+                for (packet, v) in msgs.read().unwrap().packets.iter().zip(verifies.iter()) {
```
Comment (garious) [inline src/accountant_skel.rs]: Cyclomatic complexity getting too high. Can you break this function up?

**Comment by garious** [comment]
Can you rebase?

File: src/transaction.rs [hunk 12]
```diff
@@ -7,24 +7,31 @@ use plan::{Condition, Payment, Plan};
...
         let from = from_keypair.pubkey();
         let plan = Plan::Pay(Payment { tokens, to });
         let mut tr = Transaction {
-            from,
-            plan,
-            tokens,
-            last_id,
             sig: Signature::default(),
+            data: Signed {
+                from,
```
Comment (garious) [inline src/transaction.rs]: `from` probably shouldn't be in there, eh?

Comment (garious) [reply src/transaction.rs]: Fixing this here: https://github.com/solana-labs/solana/pull/100/commits/6fec8fad573623a23b2cf6a6cd66a9f80e542be1

File: src/accountant_skel.rs [hunk 13]
```diff
@@ -104,46 +111,87 @@ impl<W: Write + Send + 'static> AccountantSkel<W> {
...
+        for x in &p.packets {
+            let rsp_addr = x.meta.addr();
+            let sz = x.meta.size;
+            if let Ok(req) = deserialize(&x.data[0..sz]) {
+                r.push(Some((req, rsp_addr)));
+            } else {
+                r.push(None);
+            }
+        }
+        r
```
Comment (garious) [inline src/accountant_skel.rs]: Where's the parallelization?

You can probably do something like this:
```rust
p.packets.par_iter().map(|x|  deserialize(&x.data[0..x.meta.size]).ok().map(|r| (r, x.meta.addr()))
```

But thinking you should leave that `Result` in the return value. That error message could be useful, especially for debugging. So:

```rust
p.packets.par_iter().map(|x|  deserialize(&x.data[0..x.meta.size]).map(|r| (r, x.meta.addr()))
```

File: src/bin/client-demo.rs [hunk 14]
```diff
@@ -13,22 +13,25 @@ use std::thread::sleep;
 use std::time::{Duration, Instant};
 
 fn main() {
-    let addr = "127.0.0.1:8000";
-    let send_addr = "127.0.0.1:8001";
+    let addr = "127.0.0.1:9000";
```
Comment (garious) [inline src/bin/client-demo.rs]: Can you find another port? This is the same port the skel test uses. Using it here means you can't leave `cargo watch` running while playing with the demo.

File: src/bin/client-demo.rs [hunk 15]
```diff
@@ -13,22 +13,25 @@ use std::thread::sleep;
...
+    let send_addr = "127.0.0.1:9001";
```
Comment (garious) [inline src/bin/client-demo.rs]: Not sure what I was thinking with that name `send_addr`. Can you rename that to `reply_addr` or `response_addr`? Also, that port can probably be `0`, eh?

File: src/accountant_skel.rs [hunk 16]
```diff
@@ -44,14 +47,14 @@ impl Request {
     /// Verify the request is valid.
     pub fn verify(&self) -> bool {
         match *self {
-            Request::Transaction(ref tr) => tr.verify(),
+            Request::Transaction(ref tr) => tr.plan_verify(),
```
Comment (garious) [inline src/accountant_skel.rs]: Please rename this to `verify_plan`

**Comment by garious** [comment]
Rebased and merged here a7f59ef3c13199fd5a9a30481f95bd11fd750191

**Pull request closed without merging**